sync/threadgroup.go

```diff
@@ -53,46 +46,125 @@ func (tg *ThreadGroup) IsStopped() bool {
 func (tg *ThreadGroup) Add() error {
 	tg.mu.Lock()
 	defer tg.mu.Unlock()
-	if tg.IsStopped() {
+
+	if tg.isStopped() {
 		return ErrStopped
 	}
 	tg.wg.Add(1)
 	return nil
 }
 
+// AddPermanent increments the ThreadGroup counter. Unlike Add, the thread
+// group will not wait for a call to 'Done' from this thread group during a
+// pause.
+func (tg *ThreadGroup) AddPermanent() error {
+	tg.mu.Lock()
+	defer tg.mu.Unlock()
+
+	if tg.isStopped() {
+		return ErrStopped
+	}
+	tg.wgPerm.Add(1)
+	return nil
+}
+
+// AfterStop adds a function to the ThreadGroup's stopFns set. Members of the
+// set will be called when Stop is called, in reverse order. If the ThreadGroup
+// is already stopped, the function will be called immediately.
+func (tg *ThreadGroup) AfterStop(fn func()) {
+	tg.mu.Lock()
+	defer tg.mu.Unlock()
+
+	if tg.isStopped() {
+		fn()
+		return
+	}
+	tg.stopFns = append(tg.stopFns, fn)
+}
+
+// BeforeStop will call a function during the 'Stop' call, but before waiting
+// for all other threads to complete.
+func (tg *ThreadGroup) BeforeStop(fn func()) {
+	tg.mu.Lock()
+	defer tg.mu.Unlock()
+
+	if tg.isStopped() {
+		fn()
+		return
+	}
+	tg.beforeStopFns = append(tg.beforeStopFns, fn)
+}
+
```

Comment:
`BeforeWait`, `AfterWait`, `beforeWaitFns`, and `afterWaitFns` might be better names, since the `stopChan` is closed before the `beforeStopFns` are called.